src/backend/cuda/copy.cpp

```diff
@@ -22,87 +22,74 @@ using common::is_complex;
 namespace cuda {
 
 template<typename T>
-void copyData(T *dst, const Array<T> &src) {
-    if (src.elements() == 0) { return; }
-
-    // FIXME: Merge this with copyArray
-    src.eval();
-
-    Array<T> out = src;
-    const T *ptr = NULL;
-
-    if (src.isLinear() ||  // No offsets, No strides
-        src.ndims() == 1   // Simple offset, no strides.
-    ) {
-        // A.get() gets data with offsets
-        ptr = src.get();
-    } else {
-        // FIXME: Think about implementing eval
-        out = copyArray(src);
-        ptr = out.get();
+void copyData(T *data, const Array<T> &src) {
+    if (src.elements() > 0) {
+        Array<T> lin = src.isReady() && src.isLinear() ? src : copyArray(src);
+        // out is now guaranteed linear
+        auto stream = cuda::getActiveStream();
+        CUDA_CHECK(cudaMemcpyAsync(data, lin.get(), lin.elements() * sizeof(T),
+                                   cudaMemcpyDeviceToHost, stream));
+        CUDA_CHECK(cudaStreamSynchronize(stream));
     }
-
-    auto stream = cuda::getActiveStream();
-    CUDA_CHECK(cudaMemcpyAsync(dst, ptr, src.elements() * sizeof(T),
-                               cudaMemcpyDeviceToHost, stream));
-    CUDA_CHECK(cudaStreamSynchronize(stream));
 }
 
 template<typename T>
 Array<T> copyArray(const Array<T> &src) {
     Array<T> out = createEmptyArray<T>(src.dims());
-    if (src.elements() == 0) { return out; }
-
-    if (src.isLinear()) {
-        CUDA_CHECK(
-            cudaMemcpyAsync(out.get(), src.get(), src.elements() * sizeof(T),
-                            cudaMemcpyDeviceToDevice, cuda::getActiveStream()));
-    } else {
+    if (src.isReady()) {
         kernel::memcopy<T>(out, src, src.ndims());
+    } else {
+        Param<T> info(out.get(), src.dims().dims, src.strides().dims);
+        evalNodes(info, src.getNode().get());
```

Comment:
Nice! This is really clever.

```suggestion
        evalNodes(out, src.getNode().get());
```
`Array<T>` objects implicity convert to `Param<T>`

Reply:
Yep, the src parameters are copied by the `out =  CreateEmptyArray<T>(src.dims())` into out.

Updated.

Reply:
I did NOT succeed in doing something similar in OPENCL.
The compiler would not convert and was confused between Param and vector<Param>...